Mathlib/Algebra/Lie/Derivation/Basic.lean

```diff
@@ -0,0 +1,349 @@
+    [LieRingModule L α] : Prop where
+  /-- `•` and `⁅⬝, ⬝⁆`  are left commutative -/
+  smul_bracket_comm : ∀ (s : S) (l : L) (a : α), s • ⁅l, a⁆ = ⁅l, s • a⁆
+
+variable {S T : Type*}
+variable [Monoid S] [DistribMulAction S M] [SMulCommClass R S M] [SMulBracketCommClass S L M]
+variable [Monoid T] [DistribMulAction T M] [SMulCommClass R T M] [SMulBracketCommClass T L M]
+
+instance : SMul S (LieDerivation R L M) :=
+  ⟨fun r D =>
+    { toLinearMap := r • D
+      leibniz' := fun a b => by simp only [LinearMap.smul_apply, coeFn_coe, apply_lie_eq_sub,
+        smul_sub, SMulBracketCommClass.smul_bracket_comm] }⟩
+
+@[simp]
+theorem coe_smul (r : S) (D : LieDerivation R L M) : ⇑(r • D) = r • ⇑D :=
+  rfl
+
+@[simp]
+theorem coe_smul_linearMap (r : S) (D : LieDerivation R L M) : ↑(r • D) = r • (D : L →ₗ[R] M) :=
+  rfl
+
+theorem smul_apply (r : S) (D : LieDerivation R L M) : (r • D) a = r • D a :=
+  rfl
+
+instance : SMulBracketCommClass ℕ L M := ⟨fun s l a => (lie_nsmul l a s).symm⟩
+
+instance : SMulBracketCommClass ℤ L M := ⟨fun s l a => (lie_zsmul l a s).symm⟩
+
+instance : AddCommMonoid (LieDerivation R L M) :=
+  coe_injective.addCommMonoid _ coe_zero coe_add fun _ _ => rfl
+
+/-- `coe_fn` as an `AddMonoidHom`. -/
+def coeFnAddMonoidHom : LieDerivation R L M →+ L → M where
+  toFun := (↑)
+  map_zero' := coe_zero
+  map_add' := coe_add
+
+instance : DistribMulAction S (LieDerivation R L M) :=
+  Function.Injective.distribMulAction coeFnAddMonoidHom coe_injective coe_smul
+
+instance [SMul S T] [IsScalarTower S T M] : IsScalarTower S T (LieDerivation R L M) :=
+  ⟨fun _ _ _ => ext fun _ => smul_assoc _ _ _⟩
+
+instance [SMulCommClass S T M] : SMulCommClass S T (LieDerivation R L M) :=
+  ⟨fun _ _ _ => ext fun _ => smul_comm _ _ _⟩
+
+end Scalar
+
+instance instModule {S : Type*} [Semiring S] [Module S M] [SMulCommClass R S M]
+    [SMulBracketCommClass S L M] : Module S (LieDerivation R L M) :=
+  Function.Injective.module S coeFnAddMonoidHom coe_injective coe_smul
+
+section RestrictScalars
+
+variable {S : Type*} [CommRing S]
+variable [LieAlgebra S L] [Module S M] [LieModule S L M]
+variable [LinearMap.CompatibleSMul L M R S]
+variable (R)
+
+/-- If `L` is both a Lie `R`-algebra and a Lie `S`-algebra, `M` is both an `R`-module and an
+`S`-module, then a Lie `S`-derivation `L → M`, which is `R`-linear is also a Lie `R`-derivation. -/
+protected def restrictScalars (d : LieDerivation S L M) : LieDerivation R L M where
+  leibniz' := d.leibniz'
+  toLinearMap := d.toLinearMap.restrictScalars R
+
+end RestrictScalars
+
+end
+
+section
+
+variable {R : Type*} [CommRing R]
+variable {L : Type*} [LieRing L] [LieAlgebra R L]
+variable {M : Type*} [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable (D : LieDerivation R L M) (a b : L)
+
+protected theorem map_neg : D (-a) = -D a :=
+  map_neg D a
+
+protected theorem map_sub : D (a - b) = D a - D b :=
+  map_sub D a b
+
+instance : Neg (LieDerivation R L M) :=
```

Comment:
In the `Derivation` theory we have the possibility that the scalars `R` are only a `Semiring` and so the file is split up into sections according to whether or not additive inverses exists. I don't think we should do that here since in Lie theory we have simplified and always demanded an additive inverse (i.e., we always have `CommRing R` and never `CommSemiring R`).

With that in mind I think the `Add (LieDerivation R L M)`, `Neg (LieDerivation R L M)`, and `Sub (LieDerivation R L M)` instances (and their API) should all be adjacent to each other, and in the same section.

Reply:
Thanks for the explanation. I fixed it in https://github.com/leanprover-community/mathlib4/pull/11790/commits/321b0fdec3e08750203325fefd1d50caeef0ec8b.

I also removed the `AddCommMonoid` instance since we have and `AddCommGroup` one. It must however stay after the scalar multiplication section because it somehow requires a multiplicative instance by integers.